### PR TITLE
Download files directly from the server

### LINKED ISSUE #817: download of non-text files fails
It is similar to issue #545 which is already resolved. However, when I download non-text files such as png files, downloaded file is damaged.

### PR DESCRIPTION
**Pull request by jasongrout**
Fixes #817.

### CODE REVIEW

**Comment by jasongrout** [comment]
It's too bad that the contents API doesn't have a way of getting a URL for a file. Here, we have to download the entire file to the browser's memory, and then tell the browser to put it on disk. If the contents api instead had a way of fetching a server path for the file, we could just point to that.

The current notebook server has the convention of the `/files/` path, but I don't think that's encoded in the API, so I'm hesitant to use it. @minrk - thoughts?

**Comment by blink1073** [comment]
cf https://github.com/jupyter/jupyter-js-services/blob/master/src/contents.ts#L197

**Comment by jasongrout** [comment]
Oh hey, nice!

**Comment by jasongrout** [comment]
This should work much better.

**Comment by ellisonbg** [comment]
Cool, I will test this now...

File: src/filebrowser/model.ts [hunk 1]
```diff
@@ -208,19 +208,13 @@ class FileBrowserModel implements IDisposable {
...
-      element.setAttribute('download', contents.name);
-      element.click();
-      return contents;
-    });
+  download(path: string): void {
+    let url = this._manager.contents.getDownloadUrl(path);
+    let element = document.createElement('a');
+    element.setAttribute('href', url);
+    element.setAttribute('download', '');
+    element.click();
```
Comment (ellisonbg) [inline src/filebrowser/model.ts]: We should probably remove the element from the DOM afterwards.

**Comment by ellisonbg** [comment]
I have confirmed this this PR work. See my one comment about deleting the DOM node...

File: src/filebrowser/model.ts (hunk 1 above)
Comment (jasongrout) [inline src/filebrowser/model.ts]: It's never added to the DOM. It should be garbage-collected after the function exits.

Comment (ellisonbg) [inline src/filebrowser/model.ts]: Ahh, missed that. Ready for merging then.

**Pull request merged**